catkin_tools/make_jobserver.py

```diff
@@ -0,0 +1,233 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+from multiprocessing import cpu_count
+from tempfile import mkstemp
+from termios import FIONREAD
+
+import array
+import fcntl
+import os
+import psutil
+import subprocess
+import time
+
+from .common import log
+
+JOBSERVER_SUPPORT_MAKEFILE = b'''
+all:
+\techo $(MAKEFLAGS) | grep -- '--jobserver-fds'
+'''
+
+
+class _MakeJobServer:
+    """
+    This class implements a GNU make job server.
+    """
+
+    # Singleton jobserver
+    _singleton = None
+
+    def __init__(self, num_jobs=None, max_load=None, max_mem=None):
+        """
+        :param num_jobs: the maximum number of jobs available
+        :param max_load: do not dispatch additional jobs if this system load
+        value is exceeded
+        :param max_mem: do not dispatch additional jobs if system physical
+        memory usage exceeds this value
+        """
+
+        assert(_MakeJobServer._singleton is None)
+
+        if not num_jobs:
+            try:
+                num_jobs = cpu_count()
+            except NotImplementedError:
+                log('@{yf}WARNING: Failed to determine the cpu_count, falling back to 1 jobs as the default.@|')
+                num_jobs = 1
+        else:
+            num_jobs = int(num_jobs)
+
+        self.num_jobs = num_jobs
+        self.max_load = max_load
+        self.max_mem = max_mem
+        self.job_pipe = os.pipe()
+
+        # Initialize the pipe with num_jobs tokens
+        for i in range(num_jobs):
+            os.write(self.job_pipe[1], b'+')
+
+    def _obtain(self):
+        """
+        Obtain a job server token. Be sure to call release() to avoid
+        deadlocks.
+        """
+
+        while True:
+            # make sure we're observing load maximums
+            if self.max_load is not None:
+                try:
+                    max_load = 8.0
+                    load = os.getloadavg()
+                    if jobserver_running_jobs() > 0 and load[1] > self.max_load:
+                        time.sleep(0.01)
+                        continue
+                except NotImplementedError:
+                    pass
+
+            # make sure we're observing memory maximum
+            if self.max_mem is not None:
+                mem_usage = psutil.phymem_usage()
+                if jobserver_running_jobs() > 0 and mem_usage.percent > self.max_mem:
+                    time.sleep(0.01)
+                    continue
+
```

Comment:
If both `self.max_load` and `self.max_mem` are set, you probably want to wait for both conditions in a single loop, right?